lib/routes/utils/signin.js

```diff
@@ -0,0 +1,485 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+'use strict'
+
+const isA = require('joi')
+const validators = require('../validators')
+const P = require('../../promise')
+const butil = require('../../crypto/butil')
+const error = require('../../error')
+const requestHelper = require('./request_helper')
+
+const BASE_36 = validators.BASE_36
+
+// An arbitrary, but very generous, limit on the number of active sessions.
+// Currently only for metrics purposes, not enforced.
+const MAX_ACTIVE_SESSIONS = 200
+
+module.exports = (log, config, customs, db, mailer)  => {
+
+  const unblockCodeLifetime = config.signinUnblock && config.signinUnblock.codeLifetime || 0
+  const unblockCodeLen = config.signinUnblock && config.signinUnblock.codeLength || 8
+
+  return {
+
+    validators: {
+      UNBLOCK_CODE: isA.string().regex(BASE_36).length(unblockCodeLen).optional()
+    },
+
+    /**
+     * Check if the password a user entered matches the one on
+     * file for the account. If it does not, flag the account with
+     * customs. Higher level code will take care of
+     * returning an error to the user.
+     */
+    checkPassword(accountRecord, password, clientAddress) {
+      if (butil.buffersAreEqual(accountRecord.authSalt, butil.ONES)) {
+        return customs.flag(clientAddress, {
+          email: accountRecord.email,
+          errno: error.ERRNO.ACCOUNT_RESET
+        }).then(() => {
+          throw error.mustResetAccount(accountRecord.email)
+        })
+      }
+      return password.verifyHash()
+        .then(verifyHash => {
+          return db.checkPassword(accountRecord.uid, verifyHash)
+        })
+        .then(match => {
+          if (match) {
+            return match
+          }
+          return customs.flag(clientAddress, {
+            email: accountRecord.email,
+            errno: error.ERRNO.INCORRECT_PASSWORD
+          }).then(() => {
+            match
```

Comment:
What is your intention here? If it's to return `match`, you're not doing it because of the braces. Either lose those or add a `return` (or ignore me if I've misunderstood).

Reply:
And we should add an appropriate assertion to the local tests too, given that this apparently wasn't causing anything to fail.

Reply:
Whoops.  I think this is indeed trying to return `match`, but at this point `match` is known to be falsey anyway, so accidentally returning `undefined` preserves the intended externally-visible behaviour and allows the tests to pass.